learn/building_programs/project_make.md

```diff
@@ -0,0 +1,452 @@
+
+# Create lists of the build artefacts in this project
+OBJS := $(addsuffix .o, $(SRCS))
+TEST_OBJS := $(addsuffix .o, $(TEST_SRCS))
+LIB := $(patsubst %, lib%.a, $(NAME))
+TEST_EXE := $(patsubst %.f90, %.exe, $(TEST_SRCS))
+
+# Declare all public targets
+.PHONY: all clean
+all: $(LIB) $(TEST_EXE)
+
+# Create the static library from the object files
+$(LIB): $(OBJS)
+	$(AR) $@ $^
+
+# Link the test executables
+$(TEST_EXE): %.exe: %.f90.o $(LIB)
+	$(LD) -o $@ $^
+
+# Create object files from Fortran source
+$(OBJS) $(TEST_OBJS): %.o: %
+	$(FC) -c -o $@ $<
+
+# Define all module interdependencies
+csv_kinds.mod := src/csv_kinds.f90.o
+csv_module.mod := src/csv_module.F90.o
+csv_parameters.mod := src/csv_parameters.f90.o
+csv_utilities.mod := src/csv_utilities.f90.o
+src/csv_module.F90.o: $(csv_utilities.mod)
+src/csv_module.F90.o: $(csv_kinds.mod)
+src/csv_module.F90.o: $(csv_parameters.mod)
+src/csv_parameters.f90.o: $(csv_kinds.mod)
+src/csv_utilities.f90.o: $(csv_kinds.mod)
+src/csv_utilities.f90.o: $(csv_parameters.mod)
+src/tests/csv_read_test.f90.o: $(csv_module.mod)
+src/tests/csv_test.f90.o: $(csv_module.mod)
+src/tests/csv_write_test.f90.o: $(csv_module.mod)
+
+# Cleanup, filter to avoid removing source code by accident
+clean:
+	$(RM) $(filter %.o, $(OBJS) $(TEST_OBJS)) $(filter %.exe, $(TEST_EXE)) $(LIB) $(wildcard *.mod)
+```
+
+Invoking ``make`` should build the static library and the test executables as
+expected:
+
+    gfortran -c -o src/csv_kinds.f90.o src/csv_kinds.f90
+    gfortran -c -o src/csv_parameters.f90.o src/csv_parameters.f90
+    gfortran -c -o src/csv_utilities.f90.o src/csv_utilities.f90
+    gfortran -c -o src/csv_module.F90.o src/csv_module.F90
+    ar rcs libcsv.a src/csv_kinds.f90.o src/csv_module.F90.o src/csv_parameters.f90.o src/csv_utilities.f90.o
+    gfortran -c -o src/tests/csv_read_test.f90.o src/tests/csv_read_test.f90
+    gfortran -o src/tests/csv_read_test.exe src/tests/csv_read_test.f90.o libcsv.a
+    gfortran -c -o src/tests/csv_test.f90.o src/tests/csv_test.f90
+    gfortran -o src/tests/csv_test.exe src/tests/csv_test.f90.o libcsv.a
+    gfortran -c -o src/tests/csv_write_test.f90.o src/tests/csv_write_test.f90
+    gfortran -o src/tests/csv_write_test.exe src/tests/csv_write_test.f90.o libcsv.a
+
+There are a few things to note there, a ``make`` build usually interlaces the
+build artifacts and the source code, unless you put extra effort into implementing
+a build directory.
+Also, right now the the source files and dependencies are specified explicitly,
+which results in several additional lines even for such a simple project.
+
+
+## Automatically generated dependencies
+
+The main drawback of ``make`` for Fortran is the missing capability to
+determine module dependencies. This is usually solved by either adding those
+by hand or automatically scanning the source code with an external tool.
+Some compilers also offer to generate dependencies in ``make`` format.
```

Comment:
I didn't know this. Which compilers can do this?

Perhaps we can add a few examples in parentheses, i.e. "Some compilers (e.g. this compiler or that compiler)"

Reply:
`ifort -gen-deps` is one example, about I never used it so far.

Reply:
```suggestion
Some compilers (like the Intel Fortran compiler) also offer to generate dependencies in ``make`` format.
```